### PR TITLE
Dsl2 conversion

### PR DESCRIPTION
**Pull request by mahesh-panchal**
Please check this works for you as well. It worked for me on Uppmax.

### CODE REVIEW

**Comment by matrulda** [comment]
Nice! Will try to have a look later this week. I'm back in the office on Thursday.

**Comment by matrulda** [comment]
I was able to start a test-run. First thing I noticed, we want the dev profile to use the containers too.

**Comment by mahesh-panchal** [comment]
That's simple enough. Use:
```
nextflow run -c custom.config -profile snpseq,dev main.nf
```
And it will use the containers from the snpseq profile.

**Comment by mahesh-panchal** [comment]
Alternatively, we separate the singularity components into a separate profile called `singularity` and then use it like so:
```
nextflow run -c custom.config -profile snpseq,singularity main.nf
nextflow run -c custom.config -profile dev,singularity main.nf
```

**Comment by matrulda** [comment]
> Alternatively, we separate the singularity components into a separate profile called `singularity` and then use it like so:
> 
> ```
> nextflow run -c custom.config -profile snpseq,singularity main.nf
> nextflow run -c custom.config -profile dev,singularity main.nf
> ```

Oh, totally forgot that you can use multiple profiles. I think that solution would be pretty clean ☝️ However, I don't really see a case were we don't want to use the containers. My suggestion would be to specify the container info on a level that applies to all profiles. What do you think?

**Comment by mahesh-panchal** [comment]
It's possible to make settings globally, but the documentation warns against it when using profiles. I'm not quite clear on the dangers of it. 

I'll make a singularity profile, which should be used all the time then. 

```
nextflow run -c custom.config -profile snpseq,singularity main.nf
nextflow run -c custom.config -profile dev,singularity main.nf
nextflow run -c custom.config -profile irma,singularity main.nf
```

File: README.md [hunk 1]
```diff
@@ -1,28 +1,58 @@
...
-
-# Pre-requisites
+## Pre-requisites
 You need to:
-  - install Nextflow
-  - install [Singularity (version > 2.6)](https://singularity.lbl.gov/install-linux#adding-the-mirror-and-installing)
-  - Have a `.genologicsrc` file configured for the Clarity LIMS db you want to use
+  - install Nextflow (e.g. using conda `conda create -n nextflow-env nextflow` or downloading from [nextflow.io](https://www.nextflow.io/)).
+  - install [Singularity (version > 2.6)](https://singularity.lbl.gov/install-linux#adding-the-mirror-and-installing).
+  - Have a `.genologicsrc` file configured for the Clarity LIMS db you want to use.
```
Comment (matrulda) [inline README.md]: We have removed the Clarity plug-in, since we don't use it. Hence, this line can be removed.

File: config/fastq_screen.conf [hunk 2]
```diff
@@ -3,14 +3,14 @@
...
-## If the bowtie binary is not in your PATH then you can 
+## If the bowtie binary is not in your PATH then you can
 ## set this value to tell the program where to find it.
 ## Uncomment the line below and set the appropriate location
 ##
 
 #BOWTIE	/usr/local/bin/bowtie/bowtie
 BOWTIE2 /usr/local/bin/bowtie2
-
+#BOWTIE2 /opt/conda/envs/fastq_screen/bin/bowtie2
```
Comment (matrulda) [inline config/fastq_screen.conf]: Is this change needed?

File: images/Singularity.fastq-screen-0.13.0 [hunk 3]
```diff
@@ -0,0 +1,23 @@
+bootstrap: docker
+from: nfcore/base:latest
```
Comment (matrulda) [inline images/Singularity.fastq-screen-0.13.0]: This Singularity image is not needed as far as I can see?

File: nextflow.config [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    mainScript = 'main.nf'
+    nextflowVersion = '>=19.10.0'
+    version = '1.0'
+}
+
+profiles {
+
+    debug { process.beforeScript = 'env'}
+
+    dev {
```
Comment (matrulda) [inline nextflow.config:15]: Error strategy should be set to ignore for `get_QC_thresholds` in `dev` profile too.

File: nextflow.config [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                container = "shub://Molmed/summary-report-development:checkqc"
+            }
+            withName: 'get_metadata' {
+                container = "shub://Molmed/summary-report-development:checkqc"
+            }
+            withName: 'interop_summary' {
+                container = "docker://quay.io/biocontainers/illumina-interop:1.0.25--0"
+            }
+            withName: 'multiqc_per_flowcell' {
+                container = "shub://Molmed/summary-report-development:multiqc"
```
Comment (matrulda) [inline nextflow.config]: Since we dropped Clarity LIMS support we can use the official image: `docker://ewels/multiqc:v1.7`

File: nextflow.config [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            }
+            withName: 'multiqc_per_project' {
+                container = "shub://Molmed/summary-report-development:multiqc"
```
Comment (matrulda) [inline nextflow.config]: Same here: `docker://ewels/multiqc:v1.7`

File: main.nf [hunk 7]
```diff
@@ -1,234 +1,292 @@
...
 
     input:
-    set val(project), file(fastq_file) from input_fastqscreen
-    file config_dir from config_dir
-    file db from fastq_screen_db
+    path runfolder
 
     output:
-    set val(project), file("*_screen.{txt,html}") into fastq_screen_results
+    path "qc_thresholds.yaml"
```
Comment (matrulda) [inline main.nf]: This output should be optional, otherwise the flowcell report won't be generated. I solved it like this in current master (It's not shown here due to the merge conflict) :   `file("qc_thresholds.yaml") optional true into qc_thresholds_result`

File: main.nf [hunk 8]
```diff
@@ -1,234 +1,292 @@
...
+def get_project_and_reads(run_folder) {
+
+    Channel
+        .fromPath("${run_folder}/${params.bcl2fastq_outdir}/**.fastq.gz", maxDepth: 10 )
+        .filter( ~/.*_[^I]\d_001\.fastq\.gz$/ )
+        .ifEmpty { "Error: No fastq files found under ${run_folder}/ !\n"; exit 1 }
+        .map {
+            it.toString().indexOf('Undetermined') > 0 ?
+                ['NoProject', it] :
+                [(it.toString() =~ /^.*\/Unaligned\/([^\/]+)\/.*\.fastq\.gz$/)[0][1],it]
```
Comment (matrulda) [inline main.nf]: Unaligned should be `bcl2fastq_outdir`.

File: main.nf [hunk 9]
```diff
@@ -1,234 +1,292 @@
...
+        get_QC_thresholds(run_folder)
+        get_metadata(run_folder)
+        project_and_reads = get_project_and_reads(params.run_folder)
+        fastqc(project_and_reads)
+        fastq_screen(project_and_reads)
+        multiqc_per_flowcell( params.run_folder,
+            fastqc.out.map{ it[1] }.collect().ifEmpty([]),
+            fastq_screen.out.map{ it[1] }.collect().ifEmpty([]),
+            interop_summary.out.map{ it[1] }.collect().ifEmpty([]),
+            get_QC_thresholds.out.collect(),
```
Comment (matrulda) [inline main.nf]: I think the case where get_QC_thresholds is empty must be handled here.

File: main.nf [hunk 10]
```diff
@@ -1,234 +1,292 @@
...
 
+workflow check_run_quality {
+
+    /* Workflow Graph
+
+    Runfolder -> InterOp                                -> MultiQCPerFlowcell
+        |     -> GetQCThresholds                        -> MultiQCPerFlowcell
+        |     -> GetMetaData                            -> MultiQCPerFlowcell + MultiQCPerProject
+        |
+        \ -> "Undeterminded/<Project>" -> FastQC        -> MultiQCPerFlowcell + MultiQCPerProject
```
Comment (matrulda) [inline main.nf]: I first read "Undeterminded/<Project>" as a path and was confused, maybe another symbol can be used for "or"? Also: Undeterminded --> Undetermined

**Review by matrulda — CHANGES_REQUESTED** [review]
Finally got time to properly look into this! 
Great work! I get it to run successfully now, but no flowcell report is generated, since the qc_thresholds input is missing. Left some comments about that and some other things. Also, can you look into the merge conflicts?

**Comment by mahesh-panchal** [comment]
Just out of curiosity, why is the get_qc_thresholds failing? Should we add a flag to make it optional to run?

**Comment by mahesh-panchal** [comment]
Also can the images directory in general be removed?

File: README.md (hunk 1 above)
Comment (mahesh-panchal) [reply README.md]: Removed

File: config/fastq_screen.conf (hunk 2 above)
Comment (mahesh-panchal) [reply config/fastq_screen.conf]: Removed #BOWTIE2 and #BOWTIE.

File: images/Singularity.fastq-screen-0.13.0 (hunk 3 above)
Comment (mahesh-panchal) [reply images/Singularity.fastq-screen-0.13.0]: Removed

File: nextflow.config (hunk 4 above)
Comment (mahesh-panchal) [reply nextflow.config:15]: Combined process specific configurations into a file and used includeConfig

File: nextflow.config (hunk 5 above)
Comment (mahesh-panchal) [reply nextflow.config]: Changed to use biocontainers image

File: nextflow.config (hunk 6 above)
Comment (mahesh-panchal) [reply nextflow.config]: changed to use biocontainers image

File: main.nf (hunk 7 above)
Comment (mahesh-panchal) [reply main.nf]: It's the same syntax more or less. Added `optional true`.

File: main.nf (hunk 8 above)
Comment (mahesh-panchal) [reply main.nf]: Good catch. Changed "Unaligned" to the variable.

File: main.nf (hunk 9 above)
Comment (mahesh-panchal) [reply main.nf]: Corrected (eventually). Collect was supposed to return a List object and ifEmpty failed with a runtime error, but it seems Collect now returns a dataflow object and ifEmpty works. Only the get_qc_thresholds should ever be empty. The other channels should throw an error before getting to multiqc if nothing is found.

File: main.nf (hunk 10 above)
Comment (mahesh-panchal) [reply main.nf]: I changed this to the command-line syntax style.

**Comment by matrulda** [comment]
> Just out of curiosity, why is the get_qc_thresholds failing? Should we add a flag to make it optional to run?

The  run type and/or read length isn't be defined in the CheckQC config, which makes it fail. It happens quite a lot that we run stuff with special read lengths that we haven't specified QC criteria for. So we want it to simply skip that step if QC thresholds aren't defined. Since we are starting this pipeline automatically when a sequencing run is complete, I don't think we want to use the parameter and instead want the pipeline to handle it. But it won't hurt to have it. :) Maybe in some cases we'll use it.   

Note: I recently implemented a feature in CheckQC that allows the nearest read length to be used if the one used isn't defined the config. I was planning on adding an option to use that in the get_qc_thresholds script, but wanted the code base to be updated first.

**Comment by matrulda** [comment]
> Also can the images directory in general be removed?

I would say no, we still need the checkqc image recipe. It is automatically uploaded to singularity hub when updated.

**Comment by matrulda** [comment]
Looks great! However, now when I see the flowcell report I notice that the interop and bcl2fastq modules are missing. Is it the same for you @mahesh-panchal ? Could be something weird with my test data though.

**Comment by mahesh-panchal** [comment]
Those multiqc data are missing for me too:
```
nxf-scratch-dir r483.uppmax.uu.se:/scratch/11692497/nxf.jQxLsn0fod
[ERROR  ]          config : Config 'custom_logo' path not found, skipping (assets/UU_logo_tranp4f125px.png)
[INFO   ]         multiqc : This is MultiQC v1.8
[INFO   ]         multiqc : Template    : default
[INFO   ]         multiqc : Report title: Flowcell report for 180126_HSX122_0568_BHLFWLBBXX_small
[INFO   ]         multiqc : Searching   : /scratch/11692497/nxf.jQxLsn0fod
[INFO   ]         multiqc : Only using modules fastqc, fastq_screen, bcl2fastq, interop, custom_content
[INFO   ]  custom_content : sequencing_metadata: Found 1 sample (html)
[INFO   ]  custom_content : sequencing_metadata: Found 592 samples (html)
[INFO   ]    fastq_screen : Found 8 reports
[INFO   ]          fastqc : Found 8 reports
[INFO   ]         multiqc : Compressing plot data
[INFO   ]         multiqc : Report      : 180126_HSX122_0568_BHLFWLBBXX_small_multiqc_report.html
[INFO   ]         multiqc : Data        : 180126_HSX122_0568_BHLFWLBBXX_small_multiqc_report_data
[INFO   ]         multiqc : MultiQC complete
```
I think I know what I need to fix.

**Comment by mahesh-panchal** [comment]
So for bcl2fastq, multiqc is looking for a `Stats.json` file (https://multiqc.info/docs/#bcl2fastq), and
for InterOp, it's looking for two files, `interop/summary` and `interop/index-summary` file (https://multiqc.info/docs/#interop), which correspond to neither of the outputs from nextflow processes.
The bcl2fastq process (`get_metadata`) produces a yaml file, not json, and `interop_summary` produces one output, not two. 

How should we go about fixing this one? Change the outputs of the applications, or update multiqc search patterns?

**Comment by matrulda** [comment]
The Stats.json is available in the run folder, in `<bcl2fastq_outdir>/Stats/`, so it should be fixed by supplying the path to `bcl2fastq_outdir` as input . Didn't see this before.
For interop I'm not sure. I'm pretty sure we only get the summary file (since we only run the summary command of interop) and it did worked before. Maybe something changed in MultiQC 1.7 --> 1.8. Might be worth trying 1.7 again.

**Comment by mahesh-panchal** [comment]
So what's the purpose of the `get_metadata` process? As far as I can see it has no purpose since the data is not reported in any way. 
I would have corrected the workflow by replacing `get_metadata.out.collect())` with `path("${params.bcl2fastq_outdir}/Stats/Stats.json"))`, which would mean then the metadata is not used anywhere. 

As for InterOp, it may not need both files and reads whatever it finds from just the one. I just need to know whether that file has the data it wants, and we can fix the rest in the multiqc config / workflow.

**Comment by matrulda** [comment]
The get_metadata constructs custom content for MultiQC, adding the Sequencing Metadata sections to the reports. 
I can help look into what's wrong with the interop stuff, but kinda swamped right now. Will let you know when I've looked into it.

**Comment by mahesh-panchal** [comment]
OK. I think I understand now. 
An extra path is needed for the bcl2fastq input, rather than replace the metadata.
For interop, it may just be a matter of giving the filename to scan for in the multiqc yaml. I'll test when Uppmax is back up.

**Comment by mahesh-panchal** [comment]
OK. The reports should have all the correct info now.

**Comment by mahesh-panchal** [comment]
The code could still do with a little clean up though, like dealing with mix of global and local parameters.

**Review by matrulda — APPROVED** [review]
Great! Works for me now too. Added a commit that updates the checkqc container paths according to the renamed checkqc image (that will be published to singularity hub, when this is merged to master.)
I'm happy to merge this to master now.

**Comment by matrulda** [comment]
> The code could still do with a little clean up though, like dealing with mix of global and local parameters.

Good enough for now I think. Thank you so much for all the work. This is such a great improvement.
